src/abstractions/IRequestAdapter.cs

```diff
@@ -50,23 +50,15 @@ public interface IRequestAdapter
         /// <param name="errorMapping">The error factories mapping to use in case of a failed request.</param>
         /// <param name="cancellationToken">The <see cref="CancellationToken"/> to use for cancelling the requests.</param>
         /// <returns>The deserialized primitive response model.</returns>
-#if NET5_0_OR_GREATER
```

Comment:
while I understand the new implementation does not require type information anymore, I'd like for us to setup a validation project with trimming to ensure that it does not produce warnings anymore.

Reply:
Should the validation project be part of this PR? Is it enough to reference projects from solution without extra code?

Reply:
I think this is a valid suggestion. Would you be in a position to add this @MartinM85?

I believe we would need to follow the guidance at https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/prepare-libraries-for-trimming?pivots=dotnet-8-0#show-all-warnings-with-test-app
and create 2 projects (one targeting net6.0 and another for net8.0)

Reply:
ok, I will do it

Reply:
Errors from validation project:
1.

> HttpClientRequestAdapter.cs(302,29): Trim analysis error IL2072: Microsoft.Kiota.Http.HttpClientLibrary.HttpClientRequestAdapter.<SendPrimitiveAsync>d__21<ModelType>.MoveNext(): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.
> PublicFields' in call to 'Microsoft.Kiota.Abstractions.Helpers.EnumHelpers.GetEnumValue(Type, String)'. The return value of method 'System.Nullable.GetUnderlyingType(Type)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.

2.

> HttpClientRequestAdapter.cs(302,29): Trim analysis error IL2087: Microsoft.Kiota.Http.HttpClientLibrary.HttpClientRequestAdapter.<SendPrimitiveAsync>d__21<ModelType>.MoveNext(): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.
> PublicFields' in call to 'Microsoft.Kiota.Abstractions.Helpers.EnumHelpers.GetEnumValue(Type, String)'. The generic parameter 'ModelType' of 'Microsoft.Kiota.Http.HttpClientLibrary.HttpClientRequestAdapter.<SendPrimitiveAsync>d__21<ModelType>' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.

3.

> JsonParseNode.cs(332,25): Trim analysis error IL2087: Microsoft.Kiota.Serialization.Json.JsonParseNode.GetCollectionOfPrimitiveValues>d__23<T>.MoveNext(): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicFields' in call to 'Microsoft.Kiota.Abstractions.Helpers.EnumHelpers.GetEnumValue(Type, String)'. The generic parameter 'T' of 'Microsoft.Kiota.Serialization.Json.JsonParseNode.<GetCollectionOfPrimitiveValues>d__23<T>' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.

4.

> JsonParseNode.cs(332,25): Trim analysis error IL2072: Microsoft.Kiota.Serialization.Json.JsonParseNode.<GetCollectionOfPrimitiveValues>d__23<T>.MoveNext(): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicFields' in call to 'Microsoft.Kiota.Abstractions.Helpers.EnumHelpers.GetEnumValue(Type, String)'. The return value of method 'System.Nullable.GetUnderlyingType(Type)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.

5.

> JsonSerializationWriter.cs(443,38): Trim analysis error IL2075: Microsoft.Kiota.Serialization.Json.JsonSerializationWriter.WriteNonParsableObjectValue<T>(String, T): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicProperties' in call to 'System.Type.GetProperties()'. The return value of method 'System.Object.GetType()' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.

6.

> .nuget\packages\azure.core\1.40.0\lib\net6.0\Azure.Core.dll : error IL2104: Assembly 'Azure.Core' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries

7.

> .nuget\packages\system.memory.data\1.0.2\lib\netstandard2.0\System.Memory.Data.dll : error IL2104: Assembly 'System.Memory.Data' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries

8.

> \.nuget\packages\microsoft.net.illink.tasks\7.0.100-1.23401.1\build\Microsoft.NET.ILLink.targets(86,5): error NETSDK1144: Optimizing assemblies for size failed. Optimization can be disabled by setting the PublishTrimmed property to false.

Errors 1-4 are related to this PR. I don't think that modifying unit tests (https://github.com/microsoft/kiota-dotnet/pull/291#discussion_r1672044115) is good, but currently I've no idea how to fix it.

Error 5 can be fixed in this PR

Error 6-8 are out of the scope of this PR

@andrueastman Should be publishing of validation projects part of Azure Pipeline (ci-build.yml)?

Reply:
> @andrueastman Should be publishing of validation projects part of Azure Pipeline (ci-build.yml)?

I think adding a project here with `<TreatWarningsAsErrors>true</TreatWarningsAsErrors>` enabled should be enough as it will make the build fail. Any chance you can push the changes as well so that we may try out the changes?

> Errors 1-4 are related to this PR. 

I believe the issue is here that a function like `GetCollectionOfPrimitiveValues` will receive a type T, but then we will call `Nullable.GetUnderlyingType` to give us another type which is then passed `EnumHelpers.GetEnumValue`. So statically, the chain doesn't have the same annotations. 

Maybe we could try to move the calling of `Nullable.GetUnderlyingType` to be inside `EnumHelpers.GetEnumValue` so that the type passed is the same type annotated at the original calling function?

Reply:
Moving Nullable.GetUnderlyingType to EnumHelpers.GetEnumValue resolved trimming errors. So, now only figure out how make unit tests working.

Reply:
@andrueastman Just a question. Any reason to reference ILLink.* in test projects?

![image](https://github.com/user-attachments/assets/7e728bdf-8dcb-4dd0-b6c3-7873cc6559a8)

It seems to me that ILLink is added automatically due to 

```
<IsTrimmable Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)','net5.0'))">true</IsTrimmable>
<IsAotCompatible Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)','net8.0'))">true</IsAotCompatible>
```

in Directory.Build.props.

If I set explicitly in test projects

```
<IsTrimmable>false</IsTrimmable>
<IsAotCompatible>false</IsAotCompatible>
```

Trimming warnings in test projects disappear.

Seems like settings from

```
 <PropertyGroup Condition="('$(IsTestProject)' != 'true') and ('$(IsAnalyzerProject)' != 'true')">
```

Is applied also to test projects

Reply:
That's odd. What you can probably do is add a `Directory.Build.Props` in the `src` folder and mobe the entire section in this clause there. (tests projects won't be in this folder....)
```xml
 <PropertyGroup Condition="('$(IsTestProject)' != 'true') and ('$(IsAnalyzerProject)' != 'true')">
```